proposals/blittable.md

```diff
@@ -0,0 +1,161 @@
+
+``` c#
+void Hash<T>(T value) where T : blittable struct
+{
+    using (T* p = &value) { 
+        ...
+    }
+}
+```
+
+Instead developers are forced to rewrite virtually the same code for all of their blittable types:
+
+``` c#
+void Hash(Point p) { 
+    ...
+}
+
+void Hash(Time t) { 
+    ...
+}
+```
+
+The lack of constraints here also make it impossible to have efficient conversions between streams of data and more structured types.  This is an operation that is common in networking stacks and serialization layers:
+
+``` c#
+Span<byte> Convert<T>(ref T value) where T : blittable {
+    ...
+}
+```
+
+Such routines are advantageous because they are provably safe at compile time and allocation free.  Interop authors today can not due this (even though it's at a layer where perf is critical).  Instead they need to rely on allocating routines that have expensive runtime checks to verify values are correctly blittable.
+
+## Detailed design
+[design]: #detailed-design
+
+The language will introduce a new declaration modifier named `blittable` that can be applied to `struct` definitions.  
+
+``` c#
+blittable struct Point 
+{
+    public int X;
+    public int Y;
+}
+```
+
+The compiler will enforce that the fields of such a struct definition fit one of the following categories:
+
+- `sbyte`, `byte`, `short`, `ushort`, `int`, `uint`, `long`, `ulong`, `char`, `float`, `double`, `decimal`, or `bool`.
+- Any `enum` type
+- Any pointer type which points to a `blittable` type
+- Any user defined struct explicitly declared as `blittable`
+
+Any type fitting the definition above, or any element in the list, is considered a valid `blittable` type.
+
+``` c#
+struct User
+{
+    string FirstName;
+    string LastName;
+}
+
+blittable struct ItemData
+{
+    // Error: blittable struct cannot contain field of type User which is not blittable
+    User User;
+    int Id;
+}
+```
+
+Note that a user defined struct must be explicitly declared as `blittable` in order to meet the requirements above.  This is required even if the struct otherwise meets the requirements of `blittable`. 
+
+``` c#
+struct SimplePoint
+{
+    public int X;
+    public int Y;
+}
+
+blittable struct Data
+{
+    // Error: blittable struct cannot contain field of type SimplePoint which is not blittable
+    SimplePoint Point;
+}
+```
+
+The language will also support the ability to constrain generic type parameters to be `blittable` types.  
+
+``` C#
+void M<T>(T p) where T : blittable struct
+{
+    ...
+}
+
+M<Point>(); // Ok
+M<User>();  // Error: Type User does not satisfy the blittable constraint.
+```
+
+One of the primary motivations for `blittable` structs is ease of interop.  As such it's imperative that such a type have it's field ordered in a sequential, or explicit, layout.  An auto layout of fields makes it impossible to reliably interop the data.  
```

Comment:
Yeah, rule of thumb I say is if it has a apostrophe, read it in your head as "it is" and see it if makes sense.